x-pack/platform/plugins/shared/fleet/server/services/package_policy.ts

```diff
@@ -1702,47 +1775,173 @@ class PackagePolicyClientImpl implements PackagePolicyClient {
-    packagePolicy?: PackagePolicy,
     pkgVersion?: string
   ): Promise<UpgradePackagePolicyResponse> {
+    // Bulk upgrade packages policies
+    // 1. bulk get policies
+    // 2. get pkg version if not specificed based on installed version
+    // 3. for each policies get upgraded
+    // 4. bulkUpdate policies
     return runWithCache(async () => {
       const result: UpgradePackagePolicyResponse = [];
+      for (const chunkedIds of chunk(MAX_CONCURRENT_AGENT_POLICIES_OPERATIONS, ids)) {
+        const packagePolicies = await this.getByIDs(soClient, chunkedIds, {
+          ignoreMissing: true,
+        });
+        const packagePoliciesById = indexBy('id', packagePolicies);
+
+        const packageVersionMap = new Map<string, string>();
+
+        if (!pkgVersion) {
+          const packageNames =
+            packagePolicies
+              ?.map((policy) => policy.package?.name)
+              .filter((packageName): packageName is string => typeof packageName !== 'undefined') ??
+            [];
+
+          await pMap(
+            uniq(packageNames),
+            async (pkgName) => {
+              const installation = await getInstallation({
+                savedObjectsClient: soClient,
+                pkgName,
+              });
 
-      for (const id of ids) {
-        try {
-          const {
-            packagePolicy: currentPackagePolicy,
-            packageInfo,
-            experimentalDataStreamFeatures,
-          } = await this.getUpgradePackagePolicyInfo(soClient, id, packagePolicy, pkgVersion);
-
-          if (currentPackagePolicy.is_managed && !options?.force) {
-            throw new PackagePolicyRestrictionRelatedError(`Cannot upgrade package policy ${id}`);
+              if (installation) {
+                packageVersionMap.set(pkgName, installation.version);
+              }
+            },
+            {
+              concurrency: MAX_CONCURRENT_AGENT_POLICIES_OPERATIONS_10,
+            }
+          );
+        }
+
+        const updatedPackagePolicies: Array<NewPackagePolicy & { version?: string; id: string }> =
+          [];
+        for (const id of chunkedIds) {
+          try {
+            const packagePolicy = packagePoliciesById[id];
+            if (!packagePolicy) {
+              throw new FleetNotFoundError(`Package policy ${id} not found`);
+            }
+
+            const upgradePkgVersion = pkgVersion
+              ? pkgVersion
+              : packageVersionMap.get(packagePolicy.package?.name ?? '');
+
+            const { packagePolicy: currentPackagePolicy, packageInfo } =
+              await this.getUpgradePackagePolicyInfo(
+                soClient,
+                id,
+                packagePolicy,
+                upgradePkgVersion
+              );
+
+            if (currentPackagePolicy.is_managed && !options?.force) {
+              throw new PackagePolicyRestrictionRelatedError(`Cannot upgrade package policy ${id}`);
+            }
```

Comment:
I see this same check is done in the nonbulk `upgrade()` function too. do you think we could move the check to `doUpgrade()`?

Reply:
fixed in https://github.com/elastic/kibana/pull/213587/commits/579ee9bc293ce1e3f6834cd7b3c9abfd7c93896f